job-server/src/main/scala/spark/jobserver/BaseSupervisorActor.scala

```diff
@@ -0,0 +1,260 @@
+  case object ContextAlreadyExists
+  case object NoSuchContext
+  case object ContextStopped
+}
+
+/**
+ * Abstract class that defines all the messages to be implemented
+ * in other supervisors
+ */
+abstract class BaseSupervisorActor extends InstrumentedActor {
+  import ContextSupervisor._
+  import scala.collection.JavaConverters._
+  import scala.concurrent.duration._
+
+  val config = context.system.settings.config
+  val defaultContextConfig = config.getConfig("spark.context-settings")
+
+  import context.dispatcher   // to get ExecutionContext for futures
+  import akka.pattern.ask
+
+  // This is for capturing results for ad-hoc jobs. Otherwise when ad-hoc job dies, resultActor also dies,
+  // and there is no way to retrieve results.
+  val globalResultActor = context.actorOf(Props[JobResultActor], "global-result-actor")
+
+  type SJSContext = (String, Option[ActorRef], Option[ActorRef])
+  // actor name -> (String contextActorName, Option[JobManagerActor] ref, Option[ResultActor] ref)
+  private val contexts = mutable.HashMap.empty[String, SJSContext]
+
+  def wrappedReceive: Receive = {
+    case MemberUp(member) =>
+      onMemberUp(member)
+
+    case ActorIdentity(memberActors, actorRefOpt) =>
+      onActorIdentity(memberActors, actorRefOpt)
+
+    case AddContextsFromConfig =>
+      addContextsFromConfig(config)
+
+    case ListContexts =>
+      sender ! listContexts()
+
+    case AddContext(name, contextConfig) =>
+      val originator = sender()
+      val mergedConfig = contextConfig.withFallback(defaultContextConfig)
+      if (haveContext(name)) {
+        originator ! ContextAlreadyExists
+      } else {
+        registerAndStartContext(name, mergedConfig, false)(
+          (ref) => originator ! ContextInitialized,
+          (err) => originator ! ContextInitError(err)
+        )
+      }
+
+    case StartAdHocContext(classPath, contextConfig) =>
+      val originator = sender()
+      val mergedConfig = contextConfig.withFallback(defaultContextConfig)
+
+      var contextName = ""
+      do {
+        contextName = java.util.UUID.randomUUID().toString().take(8) + "-" + classPath
+      } while (haveContext(contextName))
+      registerAndStartContext(contextName, mergedConfig, true)(
+        (ref) => originator ! getContext(contextName),
+        (err) => originator ! ContextInitError(err)
+      )
+
+    case GetResultActor(name) =>
+      if (haveContext(name)) {
+        sender ! getContext(name)._3.getOrElse(globalResultActor)
+      } else {
+        sender ! globalResultActor
+      }
+
+    case GetContext(name) =>
+      if (haveContext(name)) {
+        val actorOpt = getContext(name)._2
+        if (actorOpt.isDefined) {
+          val future = (actorOpt.get ? SparkContextStatus)(getTimeout().seconds)
+          val originator = sender
+          future.collect {
+            case SparkContextAlive => originator ! getContext(name)
+            case SparkContextDead =>
```

Comment:
note that if someone calles getContext while it's initiating before the internal sparkcontext is created, it will try to stop it (it might even stop a recently created context).